### PR TITLE
instructions to compile proxybroker to shared executable for Major platforms (Windows, Linux, and MacOS)

### PR DESCRIPTION
**Pull request by ziloka**
Instructions for #31. 

The steps for macos should be the same.
For windows it should be very similar the end of the `$(which proxybroker)` should be replaced with the location of the proxybroker script

### CODE REVIEW

**Comment by ziloka** [comment]
the line in proxybroker/utils.py was stolen from [this stackoverflow comment](https://stackoverflow.com/a/44352931)

**Comment by bluet** [comment]
@ziloka Is this for Linux/macOS or Windows also? AFAIK commands on Windows should be different, no?

It's OK if this is for Linux/macOS only, just change the PR title to clarify it. :-)

**Comment by ziloka** [comment]
It works on windows too, I just don't have the develop environment on windows to test it because I usually work on Linux distributions.

**Comment by ziloka** [comment]
> AFAIK commands on Windows should be different, no?

Yes they are different, which is why I stated that $(which proxybroker) needs to be replaced with the location of the python script

**Comment by ziloka** [comment]
I think python 3.9 build is broken for windows when compiling. Here is the [error and the fix for it](https://stackoverflow.com/questions/64497319/python-discord-py-error-could-not-build-wheels-for-multidict-yarl-which-use)

**Comment by ziloka** [comment]
Looking from [this github comment](https://github.com/Rapptz/discord.py/issues/5984#issue-731894511) it seems weird that the best solution is to not use python 3.9 on windows, or downgrade to python 3.8. I'm not sure about python 3.10

**Comment by ziloka** [comment]
Okay just to note here 
> Because the program makes a temporary folder with a unique name, you can run multiple copies of the app; they won’t interfere > with each other. However, running multiple copies is expensive in disk space because nothing is shared.

> The _MEIxxxxxx folder is not removed if the program crashes or is killed (kill -9 on Unix, killed by the Task Manager on
> Windows, “Force Quit” on macOS). Thus if your app crashes frequently, your users will lose disk space to multiple _MEIxxxxxx > temporary folders.

> It is possible to control the location of the _MEIxxxxxx folder by using the [--runtime-tmpdir](https://pyinstaller.org/en/stable
> /usage.html#cmdoption-runtime-tmpdir) command line option. The specified path is stored in the executable, and the bootloader > will create the _MEIxxxxxx folder inside of the specified folder. Please see [Defining the Extraction Location](https://pyinstaller.org/en/stable/usage.html#defining-the-extraction-location) for details.
Running the program and it crashing it suddenly will cause users to lose disk space due to pyinstaller having to extract said files

**Comment by ziloka** [comment]
Just curious, should we make automated builds to build executables for this project.

1. Just to know if the project can still be built with the changes made
2. People can just download them from github actions instead of compiling it themselves. For convenience sake?

**Comment by bluet** [comment]
I've been in pure Linux env for over 15 years and don't have a Windows environment too. Actually, I can't say I know how to use it to develop or test python programs :rofl: 

Before we can confirm if it really works on Windows through testing all the functions, we can't say "It works". People might believe our words and waste lots of time following the instructions we provide, which only get errors lol

If making it work on Windows is difficult for now, we can reduce the scope of this PR to support just Linux/macOSX first (and review/merge it first). For the Windows support, you can create another PR and spend time digging deeper. :smiley:

**Comment by bluet** [comment]
@ziloka auto-builds would be awesome. I just don't have the time (and knowledge, and passion) to figure out how to make windows builds. :smiling_face_with_tear: 

If you know how, PRs are welcome! :smiley:

**Comment by ziloka** [comment]
alright

**Comment by ziloka** [comment]
Okay so I dont think mac builds work either? https://github.com/ziloka/proxybroker2/runs/8000511714?check_suite_focus=true
RIght now it seems like Windows and Mac builds dont want to build which is kinda annoying.

**Comment by ziloka** [comment]
From my research I honestly just think we should limit this PR to Linux only. Until I have some type of solution for windows and macOS.

**Comment by ziloka** [comment]
Okay I have a solution for the other platforms and I think its just using a different tool for those platforms.

**Comment by ziloka** [comment]
Nevermind I used the wrong flag (for whatever reason), and Linux and macOS builds are fine
https://github.com/ziloka/proxybroker2/runs/8007781077?check_suite_focus=true

**Comment by ziloka** [comment]
Okay just to let you know, since I did a lot of research today, the reason why I can't make github actions build for windows is because of the project structure. It would honestly be better if there was a single python file that we wrote in side this project. Instead of generating that file and calling pyinstaller on it.

But I'm afraid to make a ton of reorganizing changes because I'm sure the project was structured like this for a reason.

**Comment by bluet** [comment]
> Okay just to let you know, since I did a lot of research today, the reason why I can't make github actions build for windows is because of the project structure.
> It would honestly be better if there was a single python file that we wrote in side this project. 
Do you mean to combine EVERYTHING into ONE file?

> Instead of generating that file and calling pyinstaller on it.
I don't quite understand what you mean "generating that file".

> But I'm afraid to make a ton of reorganizing changes because I'm sure the project was structured like this for a reason.
Split different sets of codes into different files/modules will make it cleaner and easier to maintain, and they'll also be reusable.

**Comment by bluet** [comment]
@ziloka BTW, #88 has been merged into `master` 3 days ago. Maybe you'd like to sync the changes back to your branch?

**Comment by ziloka** [comment]
Oh im sorry i mean #94

**Comment by ziloka** [comment]
> > Okay just to let you know, since I did a lot of research today, the reason why I can't make github actions build for windows is because of the project structure.
> > It would honestly be better if there was a single python file that we wrote in side this project.
> 
> Do you mean to combine EVERYTHING into ONE file?
> 
> > Instead of generating that file and calling pyinstaller on it.
> 
> I don't quite understand what you mean "generating that file".
> 
> > But I'm afraid to make a ton of reorganizing changes because I'm sure the project was structured like this for a reason.
> 
> Split different sets of codes into different files/modules will make it cleaner and easier to maintain, and they'll also be reusable.

I mean to explicitly have a command line .py file, instead of having setuptools generate that file. However im not sure if thats possible for a python application to be a library and a command line application while having a command line entrypoint that was written by programmers instead of having setup tools generate that file for us.

**Comment by ziloka** [comment]
I think its ready

**Comment by ziloka** [comment]
Alright, next time I will draft the PR instead of just like letting it be like this

File: README.md [hunk 1]
```diff
@@ -96,6 +96,31 @@ The latest development version can be installed directly from GitHub:
...
+&& cd build \
+&& pyinstaller --onefile --add-data "../proxybroker/data:data" --workpath ./tmp --distpath . --clean $(which proxybroker) \
+&& rm -rf tmp
+```
+
+The executable is now in dist/proxybroker directory
+
+Debug
+- pyi-archive_viewer ./dist/proxybroker to view the archived contents in the executable
+
```
Comment (bluet) [inline README.md:123]:
```suggestion

### Use pre-built Docker image
``` {.sourceCode .bash}
$ docker pull bluet/proxybroker2
```

```

**Review by bluet — CHANGES_REQUESTED** [review]

**Comment by bluet** [comment]
@ziloka would you like to also open another PR for the py-to-exe auto-build Actions? :rocket:

**Comment by ziloka** [comment]
Okay I will

**Comment by ziloka** [comment]
after thinking about this, i was being dumb about converting windows to exe, it just involves editing a bit more files, but I dont know if it interferes with it as a library.

**Comment by ziloka** [comment]
https://github.com/ziloka/proxybroker2/runs/8047370449?check_suite_focus=true

**Comment by bluet** [comment]
@ziloka cool.

BTW, something still need to be done in this PR, could you add WIP prefix in title?

**Comment by bluet** [comment]
And could you check the content and format are correct?

I don't think people need to install binutils etc on Windows too lol

Also there's a missing _```_ at the end of docker part in README. I did have that in my previous code suggestions but maybe due to a bug in GitHub it's missing now.

**Review by bluet — DISMISSED** [review]

**Comment by ziloka** [comment]
Okay I think I fixed it now

**Pull request merged**